src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking_TypeInference.cs

```diff
@@ -130,6 +130,158 @@ static void Main()
+            Assert.Equal(true, symbol.Type.IsNullable);
+        }
+
+        [Fact]
+        public void LocalVar_FlowAnalysis_02()
+        {
+            var source =
+@"class C
+{
+    static void F(string? s)
+    {
+        t = null;
+        var t = s;
+        t.ToString();
+    }
+}";
+
+            var comp = CreateStandardCompilation(source, parseOptions: TestOptions.Regular8);
+            // Ideally, there should be no warning for t.ToString();.
+            comp.VerifyDiagnostics(
+                // (5,9): error CS0841: Cannot use local variable 't' before it is declared
+                //         t = null;
+                Diagnostic(ErrorCode.ERR_VariableUsedBeforeDeclaration, "t").WithArguments("t").WithLocation(5, 9),
+                // (7,9): warning CS8602: Possible dereference of a null reference.
+                //         t.ToString();
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "t").WithLocation(7, 9));
+
+            var tree = comp.SyntaxTrees[0];
+            var model = comp.GetSemanticModel(tree);
+            var declarator = tree.GetRoot().DescendantNodes().OfType<VariableDeclaratorSyntax>().First();
+            var symbol = (LocalSymbol)model.GetDeclaredSymbol(declarator);
+            Assert.Equal("System.String", symbol.Type.ToTestDisplayString());
+            Assert.Equal(true, symbol.Type.IsNullable);
+        }
+
+        [Fact]
+        public void LocalVar_FlowAnalysis_03()
+        {
+            var source =
+@"class C
+{
+    static void F(string? s)
+    {
+        if (s == null)
+        {
+            return;
+        }
+        var t = s;
+        t.ToString();
+        t = null;
+    }
+}";
+
+            var comp = CreateStandardCompilation(source, parseOptions: TestOptions.Regular8);
+            comp.VerifyDiagnostics(
+                // (11,13): warning CS8600: Cannot convert null to non-nullable reference.
+                //         t = null;
+                Diagnostic(ErrorCode.WRN_NullAsNonNullable, "null").WithLocation(11, 13));
+
+            var tree = comp.SyntaxTrees[0];
+            var model = comp.GetSemanticModel(tree);
+            var declarator = tree.GetRoot().DescendantNodes().OfType<VariableDeclaratorSyntax>().First();
+            var symbol = (LocalSymbol)model.GetDeclaredSymbol(declarator);
+            Assert.Equal("System.String", symbol.Type.ToTestDisplayString());
+            // PROTOTYPE(NullableReferenceTypes): IsNullable should be inferred nullable state: false.
+            Assert.Equal(true, symbol.Type.IsNullable);
+        }
+
+        [Fact]
+        public void LocalVar_FlowAnalysis_04()
+        {
+            var source =
+@"class C
+{
+    static void F(int n)
+    {
+        string? s = string.Empty;
+        while (n-- > 0)
+        {
+            var t = s;
+            t.ToString();
+            t = null;
+            s = null;
+        }
+    }
+}";
```

Comment:
It would also be interesting to test:
```
    static void F(int n, string? s)
    {
        while (n-- > 0)
        {
            var t = s;
            t.ToString();
            t = null;
            s = "a";
        }
    }
```

Reply:
Added.